features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/components/layout/ContentAvoidingLayout.kt

```diff
@@ -0,0 +1,138 @@
+        val contentConstraints = if (shrinkContent) {
+            Constraints(minWidth = 0, maxWidth =  constraints.maxWidth - overlay.width)
+        } else {
+            Constraints(minWidth = 0, maxWidth = constraints.maxWidth)
+        }
+        val content = measurables.first().measure(contentConstraints)
+
+        var layoutWidth = content.width
+        var layoutHeight = content.height
+
+        val data = scope.data
+
+        // Horizontal padding = width of the component - width of its contents, but only if `hasPadding` is true
+        val internalContentHorizontalPadding = if (data.hasPadding) {
+            content.width - data.contentWidth
+        } else {
+            0
+        }
+
+        when {
+            // When the content + the overlay don't fit in the available max width, we need to move the overlay to a new row
+            !shrinkContent && data.nonOverlappingContentWidth + overlay.width > constraints.maxWidth -> {
+                layoutHeight += overlay.height
+            }
+            // If the content is smaller than the available max width, we can move the overlaidView to the right of the content
+            layoutWidth < constraints.maxWidth -> {
+                // If both the content and the overlay plus the padding can fit inside the current layoutWidth, there is no need to increase it
+                if (data.nonOverlappingContentWidth + internalContentHorizontalPadding + overlay.width > layoutWidth) {
+                    // Otherwise, we need to increase it by the width of the overlay + some padding adjustments
+                    layoutWidth += overlay.width + spacing.roundToPx() - internalContentHorizontalPadding / 2
+                }
+            }
+            else -> Unit
+        }
+
+        layoutWidth = max(layoutWidth, constraints.minWidth)
+        layoutHeight = max(layoutHeight, constraints.minHeight)
+
+        layout(layoutWidth, layoutHeight) {
+            content.placeRelative(0, 0)
+            overlay.placeRelative(layoutWidth - overlay.width, layoutHeight - overlay.height)
+        }
+    }
+}
+
+/**
+ * Data class to hold the content layout data.
+ * This is used to pass the data from the content to the [ContentAvoidingLayout].
+ *
+ * @param contentWidth The full width of the content in pixels.
+ * @param contentHeight The full height of the content in pixels.
+ * @param nonOverlappingContentWidth The width of the part of the content that can't overlap with the timestamp.
+ * @param nonOverlappingContentHeight The height of the part of the content that can't overlap with the timestamp.
+ * @param hasPadding Whether the content has padding or not. Defaults to `false`.
+ */
+@Suppress("DataClassShouldBeImmutable")
+data class ContentAvoidingLayoutData(
+    var contentWidth: Int = 0,
+    var contentHeight: Int = 0,
+    var nonOverlappingContentWidth: Int = contentWidth,
+    var nonOverlappingContentHeight: Int = contentHeight,
```

Comment:
`nonOverlappingContentHeight` appears to be unused, but maybe we can keep it for parity, or for future usage?

Reply:
Yes, that was my idea for now. If we see we don't need it for anything in the near future we can probably remove it.